src/main/java/com/google/devtools/build/remote/client/LogParserUtils.java

```diff
@@ -0,0 +1,155 @@
+
+/**
+ * Methods for printing log files.
+ */
+public class LogParserUtils {
+
+  private static final String DELIMETER =
+      "---------------------------------------------------------\n";
+
+  public static class ParamException extends Exception {
+
+    public ParamException(String message) {
+      super(message);
+    }
+  }
+
+  private String filename;
+
+  public LogParserUtils(String filename) {
+    this.filename = filename;
+  }
+
+  private String getFilename() throws ParamException {
+    if (filename.isEmpty()) {
+      throw new ParamException("This operation cannot be performed without specifying --grpc_log.");
+    }
+    return filename;
+  }
+
+
+  /**
+   * Attempt to find and print ExecuteResponse from the details of a log entry for a Watch call. If
+   * no Operation could be found in the Watch call responses, or an Operation was found but failed,
+   * a failure message is printed.
+   */
+  private static void printExecuteResponse(WatchDetails watch, PrintWriter out) throws IOException {
+    for (ChangeBatch cb : watch.getResponsesList()) {
+      for (Change ch : cb.getChangesList()) {
+        if (ch.getState() != State.EXISTS) {
+          continue;
+        }
+        Operation o = ch.getData().unpack(Operation.class);
+        if (o.getResultCase() == ResultCase.ERROR && o.getError().getCode() != Code.OK.value()) {
+          out.printf("Operation contained error: %s\n", o.getError().toString());
+          return;
+        } else if (o.getResultCase() == ResultCase.RESPONSE && o.getDone()) {
+          out.println("ExecuteResponse extracted:");
+          out.println(o.getResponse().unpack(ExecuteResponse.class).toString());
+          return;
+        }
+      }
+    }
+    out.println("Could not find ExecuteResponse in Watch call details.");
+  }
+
+  /**
+   * Print an individual log entry.
+   */
+  static void printLogEntry(LogEntry entry, PrintWriter out) throws IOException {
+    out.println(entry.toString());
+    if (entry.getDetails().getDetailsCase() == DetailsCase.WATCH) {
+      out.println("\nAttempted to extract ExecuteResponse from Watch call responses:");
+      printExecuteResponse(entry.getDetails().getWatch(), out);
+    }
+  }
+
+  /**
+   * Prints each entry out individually (ungrouped) and a message at the end for how many entries
+   * were printed/skipped.
+   */
+  private void printEntriesInOrder(OutputStream outStream) throws IOException, ParamException {
+    try (InputStream in = new FileInputStream(getFilename())) {
+      PrintWriter out =
+          new PrintWriter(new BufferedWriter(new OutputStreamWriter(outStream, UTF_8)), true);
+      LogEntry entry;
+      while ((entry = LogEntry.parseDelimitedFrom(in)) != null) {
+        printLogEntry(entry, out);
+        System.out.print(DELIMETER);
+      }
+    }
+  }
+
+  private void printEntriesGroupedByAction(OutputStream outStream)
+      throws IOException, ParamException {
+    ActionGrouping byAction = new ActionGrouping();
```

Comment:
It would be nice to sort the actions themselves as well, for example by the time of the first RPC.

Reply:
Agreed. I'll add it in a followup cl with unit tests for ActionGrouping.

Sorry this is a bit half-baked; I was preparing a commit train with the ShowAction changes as well, but I realized that I need this change to go in before my API v2 migration change (I've moved files around and it's a pain to resolve conflicts when I try to rebase).